File: apps/macos/Tests/WispProjectBrowserTests/NativePanelTabsTests.swift
import XCTest
@testable import WispProjectBrowser

final class NativePanelTabsTests: XCTestCase {
    func testRestoresOrderAndSanitizesOldPreferences() {
        let state = NativePanelTabs(saved: "[\"hosts\",\"files\",\"hosts\",\"future\",\"notebook\"]", selected: "future")
        XCTAssertEqual(state.open, ["hosts", "files"])
        XCTAssertEqual(state.selected, "hosts")
        XCTAssertEqual(NativePanelTabs(saved: "broken").open, NativePanelTabs.defaults)
        XCTAssertEqual(NativePanelTabs(saved: state.saved, selected: state.selected), state)
    }
    func testClosingUsesLeftNeighborAndLastCloseKeepsEmptyUntilReopened() {
        var state = NativePanelTabs(selected: "files")
        state.remove("files")
        XCTAssertEqual(state.selected, "agents")
        state.remove("artifacts")
        XCTAssertEqual(state.selected, "agents")
        state.remove("agents")
        XCTAssertEqual(state.selected, "hosts")
        state.remove("hosts")
        XCTAssertTrue(state.open.isEmpty)
        XCTAssertTrue(NativePanelTabs(saved: state.saved).open.isEmpty)
        state.reopen()
        XCTAssertEqual(state.open, NativePanelTabs.defaults)
        XCTAssertEqual(state.selected, "artifacts")
    }
    func testMoveAndShowPreserveSelectionAndNeverDuplicate() {
        var state = NativePanelTabs(selected: "agents")
        state.move("artifacts", to: "files")
        XCTAssertEqual(state.open, ["agents", "files", "artifacts", "hosts"])
        XCTAssertEqual(state.selected, "agents")
        state.move("hosts", to: "agents")
        XCTAssertEqual(state.open, ["hosts", "agents", "files", "artifacts"])
        state.remove("files"); state.show("files"); state.show("files")
        XCTAssertEqual(state.open, ["hosts", "agents", "artifacts", "files"])
        XCTAssertEqual(state.selected, "files")
        state.show("notebook")
        XCTAssertEqual(state.selected, "files")
        state.move("unknown", to: "files")
        XCTAssertEqual(state.open.count, 4)
    }
    func testSidebarFilesRevealsTheExistingPageWithoutADuplicateTab() {
        let closed = NativePanelTabs(saved: "[\"artifacts\",\"agents\",\"hosts\"]", selected: "agents", available: NativePanelTabs.all)
        let revealed = NativePanelTabs.revealFiles(saved: closed.saved, selected: closed.selected)
        XCTAssertTrue(revealed.visible)
        XCTAssertEqual(revealed.selected, "files")
        let restored = NativePanelTabs(saved: revealed.saved, selected: revealed.selected, available: NativePanelTabs.all)
        XCTAssertEqual(restored.open.filter { $0 == "files" }, ["files"])
        XCTAssertEqual(restored.selected, "files")
        let again = NativePanelTabs.revealFiles(saved: revealed.saved, selected: revealed.selected)
        let repeated = NativePanelTabs(saved: again.saved, selected: again.selected, available: NativePanelTabs.all)
        XCTAssertEqual(repeated.open.filter { $0 == "files" }, ["files"])
        XCTAssertTrue(again.visible)
    }
    func testOptionalSurfacesCanOptIntoStableRegistry() {
        var state = NativePanelTabs(available: NativePanelTabs.all)
        state.show("notebook"); state.show("highlights"); state.show("provenance"); state.show("sidechat")
        XCTAssertEqual(state.open.count, 8)
        XCTAssertEqual(NativePanelTabs(saved: state.saved, selected: state.selected, available: NativePanelTabs.all), state)
    }
}

File: apps/windows/Wisp.ProjectBrowser.ContractTests/NativePanelTabsTests.cs
using Wisp.ProjectBrowser.Contracts;

static class NativePanelTabsTests
{
    private static void Require(bool value) { if (!value) throw new InvalidOperationException("Native panel tab behavior drift"); }
    public static void Run()
    {
        var restored = new NativePanelTabs("[\"hosts\",\"files\",\"hosts\",\"future\",\"notebook\"]", "future");
        Require(restored.Open.SequenceEqual(new[] { "hosts", "files" }) && restored.Selected == "hosts");
        Require(new NativePanelTabs("broken").Open.SequenceEqual(NativePanelTabs.Defaults));
        Require(new NativePanelTabs(restored.Saved, restored.Selected).Open.SequenceEqual(restored.Open));
        foreach (var saved in new[] { "[]", "[\"agents\"]" })
        {
            var reveal = new NativePanelTabs(saved, "files", NativePanelTabs.All);
            reveal.Show("files");
            var reopened = new NativePanelTabs(reveal.Saved, reveal.Selected, NativePanelTabs.All);
            Require(reopened.Selected == "files" && reopened.Open.Count(id => id == "files") == 1);
        }
        var state = new NativePanelTabs(selected: "files");
        state.Remove("files"); Require(state.Selected == "agents");
        state.Remove("artifacts"); Require(state.Selected == "agents");
        state.Remove("agents"); Require(state.Selected == "hosts");
        state.Remove("hosts"); Require(state.Open.Count == 0);
        Require(new NativePanelTabs(state.Saved).Open.Count == 0);
        state.Reopen(); Require(state.Open.SequenceEqual(NativePanelTabs.Defaults) && state.Selected == "artifacts");
        state.Show("agents"); state.Move("artifacts", "files");
        Require(state.Open.SequenceEqual(new[] { "agents", "files", "artifacts", "hosts" }) && state.Selected == "agents");
        state.Move("hosts", "agents"); state.Remove("files"); state.Show("files"); state.Show("files");
        Require(state.Open.SequenceEqual(new[] { "hosts", "agents", "artifacts", "files" }));
        state.Show("notebook"); state.Move("unknown", "files"); Require(state.Selected == "files" && state.Open.Count == 4);
        var full = new NativePanelTabs(available: NativePanelTabs.All);
        foreach (var id in new[] { "notebook", "highlights", "provenance", "sidechat" }) full.Show(id);
        Require(full.Open.Count == 8 && new NativePanelTabs(full.Saved, full.Selected, NativePanelTabs.All).Open.SequenceEqual(full.Open));
        Console.WriteLine("Native panel tab restoration, close, reorder and optional registry tests passed.");
    }
}
